File: src/henaxel/raytracer/materials/Dielectric.java
package henaxel.raytracer.materials;

import henaxel.raytracer.node.InLink;
import henaxel.raytracer.utils.Vec3;
import henaxel.raytracer.Ray;
import java.awt.*;

public class Dielectric extends Material {
    private Texture albedo;
    private Texture refIndex;


    public Dielectric(Double refIndex) {
        this(new ConstantTexture(1.0, 1.0, 1.0), new ConstantTexture(refIndex, refIndex, refIndex));
    }
    
    public Dielectric(Texture albedo, Texture refIndex) {
        super("Dielectric",
                new InLink[] {
                        new InLink("Albedo", null),
                        new InLink("Refractive Index", null)
                }
        );
        this.albedo = albedo;
        this.refIndex = refIndex;
    }



    @Override
    public Ray scatter(Ray r, Vec3 pos, Vec3 n) {
        Vec3 outN;
        double rRatio;
        double reflectProb;
        double cosine;
        
        double rIndex = refIndex.color(0,0).getRed()/255.0; //TODO: Implement proper texture mapping for refractive index

        if (Vec3.dot(r.direction(), n) > 0) {
            outN = Vec3.multiply(n, -1);
            rRatio = rIndex ;
            cosine = rIndex * Vec3.dot(r.direction(), n) / r.direction().length();
        } else {
            outN = n;
            rRatio = 1/rIndex;
            cosine = -Vec3.dot(r.direction(), n) / r.direction().length();
        }

        Vec3 refracted = Material.refract(r.direction(), outN, rRatio);
        if(refracted != null) {
            reflectProb = Material.schlick(cosine, rIndex);
        } else {
            reflectProb = 1.0;
        }

        if(Math.random() <= reflectProb) {
            return new Ray(pos, Material.reflect(r.direction(), n));
        } else {
            return new Ray(pos, refracted);
        }
    }

    @Override
    public Ray scatter(Ray r, Vec3 pos) { return null; }

    @Override
    public Color getAlbedo(double u, double v) {
        return albedo.color(u, v);
    }
}

File: src/henaxel/raytracer/materials/Dielectric.kt
package henaxel.raytracer.materials

import henaxel.raytracer.Ray
import henaxel.raytracer.utils.Vec3
import jdk.jshell.spi.ExecutionControl
import java.awt.Color
import kotlin.random.Random

class Dielectric(var albedo: Texture, var refIndex: Texture) : Material() {

    override fun scatter(ray: Ray, pos: Vec3, normal: Vec3): Ray {

        var outN = Vec3(0,0,0)
        var rRatio = 0.0
        var reflectProb = 0.0
        var cosine = 0.0
        val rIndex = refIndex.color(0.0,0.0).red/255.0

        if ((ray.dir * normal) > 0) {
            outN = -normal
            rRatio = rIndex
            cosine = rIndex * (ray.dir * normal) / ray.dir.length()
        } else {
            outN = normal
            rRatio = 1/rIndex
            cosine = -(ray.dir * normal / ray.dir.length())
        }

        val refracted = refract(ray.dir, outN, rRatio)
        reflectProb =
            if(refracted != null)
                schlick(cosine, rIndex)
            else
                1.0

        return if (Random.nextDouble(0.0, 1.0) <= reflectProb)
            Ray(pos, reflect(ray.dir, normal))
        else
            Ray(pos, refracted!!)
    }

    // Volumetric scatter
    override fun scatter(ray: Ray, pos: Vec3): Ray { throw NotImplementedError("Volumetric scattering for dielectric material not yet implemented") }

    override fun getAlbedo(u: Double, v: Double): Color { return albedo.color(u,v) }
}